File: stencils/mwes/jl_call2_int.c
#include <stdio.h>
#include "common.h"

int
_JIT_ENTRY()
{
    PATCH_VALUE(jl_function_t *, func, _JIT_FUNC);
    PATCH_VALUE(int, _arg1, _JIT_ARG1);
    PATCH_VALUE(int, _arg2, _JIT_ARG2);
    jl_value_t *arg1 = jl_box_int32(_arg1);
    jl_value_t *arg2 = jl_box_int32(_arg2);
    jl_value_t *ret = jl_call2(func, arg1, arg2);
    int _ret = jl_unbox_int32(ret);
    return _ret;
}

File: stencils/mwes/jl_call2_int.jl
using CopyAndPatch
using JSON
using Base.Libc
using Base.Libc.Libdl

lib = dlopen(dlpath("libjulia"))
libc = dlopen(dlpath("libc.so.6"))

group = CopyAndPatch.StencilGroup(joinpath(@__DIR__, "jl_call2_int.json"))
holes = CopyAndPatch.holes(group)
bvec = CopyAndPatch.ByteVector(UInt8.(group.code.body[1]))
if !isempty(group.data.body)
    bvec_data = CopyAndPatch.ByteVector(UInt8.(group.data.body[1]))
else
    bvec_data = Any[]
end

patches = Dict{String,Any}(
    "_JIT_FUNC" => CopyAndPatch.pointer_from_function(+),
    "_JIT_ARG1" => UInt64(Int32(255)),
    "_JIT_ARG2" => UInt64(Int32(6)),
    "jl_box_int32" => dlsym(lib, :jl_box_int32),
    "jl_unbox_int32" => dlsym(lib, :jl_unbox_int32),
    "jl_call2" => dlsym(lib, :jl_call2),
)

for h in holes
    if startswith(h.symbol, ".rodata")
        p = pointer(bvec_data.d, h.addend+1)
    else
        local p = get(patches, h.symbol) do
            CopyAndPatch.TODO(h.symbol)
        end
    end
    bvec[h.offset+1] = p
end

fn = CopyAndPatch.MachineCode(bvec, Cint, ())
@show fn()
